### DISCUSSION TITLE
Global current working directory dynamic var

### DISCUSSION THREAD

**Discussion by borkdude** [babashka/babashka#1278] [Ideas]
Discussion for issue https://github.com/babashka/babashka/issues/1277

We could have a separate library and namespace for this:

```
babashka.cwd/*cwd*`
```

or we could just stick in into `babashka.fs/*cwd*`.

The idea would be that you could change the current working directory and functions in `fs` and `process` would take that into account. 

We could also patch clojure core functions, but that would maybe lead to unexpected results later on if we patch things gradually.

Alternatively, we could investigate if JNI / Graal C lets us change the current working dir.

**Comment by phmeier-nubank** [comment]
> Alternatively, we could investigate if JNI / Graal C lets us change the current working dir.

While looking tempting I wonder if other java tooling / library could be confused.

**Reply by borkdude** [reply]
Valid concern. Here is some info from the GraalVM slack: https://graalvm.slack.com/archives/CN9KSFB40/p1653659136057609
The tl;dr: JNI or Panama should work.

**Reply by borkdude** [reply]
With panama around the corner, we could wait for that maybe

**Reply by IGJoshua** [reply]
What in particular would you be waiting for with panama? It's already in incubator, are you waiting for a full release? Most likely that won't be happening till the next LTS release of JDK 21 in September of next year, at least based on my current expectations.

**Reply by IGJoshua** [reply]
If you need a consistent API to it across JVM versions you could pull in https://github.com/IGJoshua/coffi which won't have its external api changed much between versions and which will have a published release version for each jdk from 17 up.

**Comment by frenchy64** [comment]
Might work well for fs/process, but not a fan of patching Clojure functions.

Also, dynamic vars are somewhat brittle in Clojure when combined with laziness and parallelism.

**Reply by borkdude** [reply]
Agreed, we should scope the cwd thing to babashka.* libraries.
Also agreed that a dynvar can be problematic with laziness, but on the other hand, an atom can be problematic with threads, as one thread could mutate the current directory for another thread.

**Reply by frenchy64** [reply]
> Also agreed that a dynvar can be problematic with laziness, but on the other hand, an atom can be problematic with threads, as one thread could mutate the current directory for another thread.

Yes. That would not be a great implementation.

**Reply by borkdude** [reply]
So then what? ;)

**Reply by frenchy64** [reply]
I don't know. I'll try decomplecting the feature.

AFAICT this feature would make it easier to:
1. set the initial cwd for a bunch of commands
2. simultaneously change cwd between commands

Maybe 1) is covered by local bindings and file-local defn wrappers.

Maybe 2) could be covered by adding an explicit cwd atom arg to every function?

```clojure
(with-shared-cwd-atom [cwd "foo"]
  (process {:cwd-atom cwd})
  (process {:cwd-atom cwd}))
```

**Reply by borkdude** [reply]
Then you might as well just pass the arguments directly: 

``` clojure
(let [cwd "foo"]
  (fs/copy (fs/path cwd "subdir" "file.txt") (fs/path cwd))
  (process ["dude"] {:dir cwd}))
```
so maybe we just don't need this feature.

**Reply by frenchy64** [reply]
Yes, that's what I would do.

Another thought: if a library exposed a function using fs/process ops, it's not necessarily a good idea to change its cwd (maybe it's directory sensitive). It's probably better to ask the maintainer to expose an option like `:dir`.

Though I'm guessing the original motivation wasn't using libs, but making it easier to write your own scripts in a bash-like way?

**Reply by borkdude** [reply]
> but making it easier to write your own scripts in a bash-like way

Indeed. But maybe we should just not attempt that. But sometimes people find this important:

https://twitter.com/nikitonsky/status/1441126042781900809

**Reply by frenchy64** [reply]
> Indeed. But maybe we should just not attempt that. But sometimes people find this important:

Might be a good example for the bb book for such people.

```clojure
(defn process [cmd opts]
  (process/process cmd (assoc opts :dir "my-dir"))
;; or
(let [common-opts {:dir "my-dir"}]
  (process cmd1 (into common-opts {...}))
  (process cmd2 (into common-opts {...})))
```

**Reply by borkdude** [reply]
I think he already tried that, but was too much used to being able to chdir in Python :)

**Reply by frenchy64** [reply]
Hmm. Well I think `binding` would be the best way to implement this feature request if it's important. But someone from bash/python might be the most likely to shoot themselves in the foot with the subtle implications.

**Reply by borkdude** [reply]
Agreed. I might close this issue.

**Reply by IGJoshua** [reply]
`bound-fn` solves all the problems with dynamic vars and laziness, as long as people know to use them. Perhaps the problem is more with education than with dynamic variables?

**Reply by frenchy64** [reply]
Good point. I agree it's certainly possible to use correctly. Though, I think there's a cognitive overhead that should be justified by a good problem statement.

In my experience, once thread-local bindings are involved (beyond using `bound-fn`):
1. you have to know the implementation of every macro you use, or only use macros that explicitly promise binding conveyance.
2. as a library provider, you have to worry about your users pulling the rug out from underneath you. eg., what if you have your own `cwd` but someone rebinds it? or your fn cannot have its `cwd` changed, but the user expects it? (and every other corner case)
